stdlib/string.ml

```diff
@@ -100,15 +100,19 @@ let trim s =
     then bts (B.trim (bos s))
   else s
 
+let needs_escape s =
+  let n = length s in
+  let i = ref 0 in
+  let esc = ref false in
+  while !i < n && not !esc do
+    match unsafe_get s !i with
+    | '\"' | '\\' | '\000'..'\031' | '\127'.. '\255' -> esc := true
+    | _ -> incr i
+  done;
+  !esc
```

Comment:
This "while not !finished do ... done" idiom brings bad memory from Pascal.  What's wrong with the original recursion?  Remember, this is a functional language.

Reply:
Speedups come from:

  - A while-loop is still faster than tail-rec loops.
  - Not recomputing length for every character.
  - Patterns more friendly to the PM compiler (I was suprised, but the cmm code for the original code is clearly sub-optimal).

I'll post benchmarks later on.

Reply:
Not recomputing the length is fine.  Pascal style is not to be encouraged.

Reply:
Ok, so I propose:

````ocaml
let needs_escape s =
  let rec needs_escape n i =
    if i >= n then false else
      match String.unsafe_get s i with
      | '\"' | '\\' | '\n' | '\t' | '\r' | '\b' -> true
      | ' ' .. '~' -> needs_escape n (i+1)
      | _ -> true
  in
  needs_escape (String.length s) 0
`````

which gives most of the gain.

Reply:
Some very quick testing suggests that we go from about 550 Mb/s for the current version to 715 Mb/s with the version above.